[v6.2] Articles for Refactor (#6929)
gravitational/teleport: docs/pages/server-access/architecture.mdx

---
title: Teleport Server Access Architecture Overview
description: Surveying key concepts and architectural concerns when using Teleport Server Access.
h1: Teleport Server Access
---

Many administrators will consider several factors when configuring Teleport for their specific stack, architecture, and resources.

This document surveys key concepts and architectural concerns when using Teleport Server Access.

> ## Server Access security

Author: This section should probably be moved into the planned Architecture/Security article.

@benarent should we put best practices for TTL in there, into the identity management article, both, something else?

Reviewer: Maybe we can created a dedicated section that explains token TTL, and how to deal with secrets in general. We don't do a good job explaining that you can also load tokens from a file and other best practices.

Author: Great idea. I haven't gotten to these articles as much as I'd have wanted to yet but I've started to stub that out.

---
title: Teleport Server Access Architecture Overview
description: Surveying key concepts and architectural concerns when using Teleport Server Access.
h1: Teleport Server Access
---

Many administrators will consider several factors when configuring Teleport for their specific stack, architecture, and resources.

This document surveys key concepts and architectural concerns when using Teleport Server Access.

> ## Choose your SSH solution

Author: This section will likely be cut and moved into https://github.com/gravitational/teleport/pull/7043

Some of the security topics can be moved into the planned Security Architecture Guide.

</Admonition>

### Clients

Teleport is compatible with popular clients like [OpenSSH](./guides/openssh.mdx) and also provides its own [built-in SSH client](../cli-docs.mdx#tsh) (`tsh`).

By default, SSH clients can connect to a Teleport cluster through the following port (exposed through the public proxy):

| Port | Service | Description |
| - | - | - |
| 3023 | Proxy | SSH port clients connect to. A proxy will forward this connection to port `#3022` on the destination Node. |

Teleport allows for complex, multi-client, arrangments as well as unified solutions (using Teleport alone in both a dedicated SSH server and as a client through `tsh`). For example, one cluster may be accessible through OpenSSH, another through Teleport `tsh`, and another through Putty.
  
### Servers

Teleport is agnostic concerning your choice of SSH server backends and can be used to supplement existing features or serve as an intermediary between them and public clients.

A common strategy involves using Teleport Nodes in Proxy mode to get all the benefits of Audit Logging, Session Recording, and/or 2FA.

> Most backends likely expose port `22` for inbound public clients and `tsh` will connect through the specified `--port` value on command:

Reviewer: Port 22 is reserved for SSH https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers

Author: True, some folks might use a customized port number instead.

```bash
# tsh ssh into a public proxy Node using a specific port
tsh ssh --port=22 user@tele.example.com
```

## Server Access security

Security is a natural and paramount concern for any admin. And, Teleport was built to empower admins by simplifying the configuration of the often byzantine network of resources required for modern operations.

Teleport supports many security best practices out-of-the-box through its many features:

> - Role-Based Access Control (RBAC) and *The Principle of Least Privilege* (POLP) through [Role Templates](../access-controls/guides/role-templates.mdx) and [Access Request workflows](../enterprise/workflow/index.mdx).

Reviewer: TIL this acronym. https://en.wikipedia.org/wiki/Principle_of_least_privilege  It looks like it should be PoLP

This section describes additional security best practices Teleport can be easily configured to support or implement.

### Secure Administrator Workstations

A common practice at many companies is to restrict access to critical resources not only by role but also by physical (or virtual) machines as well.

According to this best practice, even if a role or credential is compromised, a breach is mitigated since only one machine can be used by that role or credential to access production resources.

Teleport supports Secure Administrator Workstations (SAW) in the following ways:

1. Access through Teleport is Teleport-to-Teleport. User access is limited to the lifespan of a valid certificate granted by the Teleport Certificate Authority. While this is a powerful security feature in its own right, it also lends itself to SAW configurations since the connection from the SAW is further restricted and mediated beyond SSH alone.
2. Teleport brings user roles and user certificates into one tool: `tctl`. This makes it easier to restrict or remove user access in the prevention or the event of a security event.

> 3. Teleport requires Two-Factor Authentication (TFA/2FA) with the option to use a physical keyfob or SSO increasing security at the SAW access point.

Reviewer: ```suggestion
3. Teleport requires Two-Factor Authentication (2FA) with the option to use U2F in the form of a hardware token or SSO increasing security at the SAW access point.
```

Once Teleport has been installed, Nodes connected, and initial users and permission created any remaining ports can be closed on both the *Bastion Host* and the protected, private, resources.

### Divided topology

Many architectures are now being designed to mitigate security events or disasters by splitting up resources and permissions into minimal groupings.

Divided topologies involve deliberately restricting traffic, filtering IP addresses, regulating permissions, and grouping resources to mitigate loss or access in the advent of a disaster or security event.

Teleport divides and groups access across resources through [Labels](../admin-guide.mdx#labeling-nodes-and-applications) and Node-based clustering. 

- Admins can dynamically modify which Nodes join or are a part of which cluster.
- Labels provide convenient groupings when querying, introspecting, or visualizing resources without modifying underlying clusters or access.

## Cluster introspection

Teleport simplifies managing and introspecting your cluster.

> ### Unified Resource Catalog

Reviewer: ```suggestion
### Resource Catalog
```

I would drop unified here, to be inline with dropping it from Unified Access Plane

> Resource admins will likely use the Teleport Unified Resource Catalog to easily manage resources by cluster and [Label](../admin-guide.mdx#labeling-nodes-and-applications).

Reviewer: ```suggestion
Resource admins will likely use the Teleport Resource Catalog to easily manage resources by cluster and [Label](../admin-guide.mdx#labeling-nodes-and-applications).
```

---------------- ------------------------------------ ------------------ ------------------------------------ 
ip-172-31-40-99  bf840c02-68f4-4fac-a9e6-30141e82d242 127.0.0.1:3022     env=example,hostname=ip-172-31-40-99 
ip-172-31-39-190 d32945f4-6e35-4318-b8f6-198d7c1fbd4c 172.31.39.190:3022                                      
```

Note the "Labels" column on the farthest side.

### Labels

Labels are permanent or temporary taxonomic handles for grouping resources. They can be used to group resources temporarily, across multiple clusters, or to define a grouping useful for querying without modifying the underlying access permissions or cluster membership:

- `tsh ssh root@env=example` returns the list of all resources with the `env=example` label specified in their `teleport.yaml` (`'env': 'example'`).
- Suppose an IP address changes, an admin can quickly find the current Node with that label since it remains unchanged.
- Suppose you need to query the SSH Node of each cluster. Give each SSH Node the same label so a single query will allow you to introspect those specific resources across all of your clusters.

> This was designed in tandem with the Unified Resource Catalog to make it easier to manage mass-scale deployments (common to many industries like IoT and manufacturing).

Reviewer: I would add a link that labels can be both dynamic and static. Dynamic tokens can be very powerful, in providing context and combined with RBAC

## Defining access request workflows

<Admonition title="tip" type="tip">
  Access Requests and Access Request workflows are now available in both the Teleport Open Source and Enterprise Editions.
</Admonition>

Granting and revoking permissions is a natural part of any system for managing identities, information, and access to resources.

Teleport provides built-in support for [requesting, renewing, granting, revoking, and denying permissions](../enterprise/workflow/index.mdx) through the same Web Interface and CLI used to access permitted resources.

## Auditing and events

Teleport provides powerful [Session Recording](../features/enhanced-session-recording.mdx) and [Audit Logging](../admin-guide.mdx#audit-log) that can be added to your existing SSH solution:

- User sessions are recorded and can be replayed. 
- They can also be shared, stored, and organized by query and triggered Events.

### Defining Events

Teleport admins can further define [Events](../admin-guide.mdx#events) to organize and trigger sending session recordings to customizable storage back-ends:

> 1. All `session.start` events could be stored on AWS S3.

Reviewer: We support a range of backends, S3, S3 Compatible and GCP File store.  I would make this section less vendor specific
